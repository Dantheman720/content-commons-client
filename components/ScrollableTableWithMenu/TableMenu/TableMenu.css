.items_menu {
  position: relative;
  display: inline-block;
  cursor: pointer;
  color: #112e51; }
  .items_menu_wrapper {
    text-align: right; }
    @media screen and (max-width: 767px) {
      .items_menu_wrapper {
        display: none !important; } }
  .items_menu.active::after {
    content: '';
    position: absolute;
    bottom: -2px;
    left: -5px;
    width: 100%;
    height: 3px;
    background-color: #112e51; }
  .items_menu [data-tablearrow]:hover {
    background-color: #0071bc;
    color: #fff; }
  .items_menu [data-tablearrow='right'].right.icon {
    margin-right: 0; }
  .items_menu_list {
    position: absolute;
    left: -5px;
    top: 100%;
    z-index: 4;
    display: none;
    width: 100%;
    padding: 5px;
    margin-top: 3px;
    border: 1px solid #ccc;
    text-align: left;
    background: #fff; }
    .items_menu_list.display {
      display: block; }
    .items_menu_list .ui.checkbox {
      display: block;
      margin-bottom: 5px; }
