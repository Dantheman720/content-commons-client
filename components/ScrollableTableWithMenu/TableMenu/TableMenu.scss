@import "../../../styles/colors.scss";

.items {
  &_menu {
    position: relative;
    display: inline-block;    
    font-size: 14px;
    color: $blue;
    cursor: pointer;

    .ui.vertical.menu {
      position: relative;
      display: inline-block;
      border: 0;
      width: auto;
      min-height: initial;
      margin: 0;
      font-size: 0.777777778rem;
      -webkit-box-shadow: none;
      box-shadow: none;

      .item {
        padding: 0.25rem 0.5rem;

        .content {
          position: absolute;
          left: 0;
          z-index: 10;
          width: 100%;
          text-align: left;
          background-color: #fff;
        }

        button {
          border: 0;
          height: 100%;
          width: 100%;
          color: $blue;
          text-align: left;
          background-color: #fff;
        }

        .grouped {
          border: 1px solid #ccc;
          border-top: 3px solid $blue;
          padding: 0.5rem;

          [data-tablemenuitem] {
            padding: 0.25rem 0;
          }
        }
      }
    }

    .ui.checkbox input.hidden + label {
      font-size: 0.777777778rem;
    }

    &_wrapper {
      text-align: right;
      @media screen and (max-width: 767px) {
        display: none !important;
      }
    }

    [data-tablearrow]:hover {
      background-color: $light-blue;
      color: #fff;
    }
    
    [data-tablearrow='right'].right.icon {
      margin-right: 0;
    }
  }
}
