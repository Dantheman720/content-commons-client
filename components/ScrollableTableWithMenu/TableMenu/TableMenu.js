/**
 *
 * TableMenu
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import {
  Accordion, Checkbox, Icon, Menu
} from 'semantic-ui-react';
import VisuallyHidden from 'components/VisuallyHidden/VisuallyHidden';
import { titleCase } from 'lib/utils';
import { isMobile, isWindowWidthLessThanOrEqualTo } from 'lib/browser';
import './TableMenu.scss';

class TableMenu extends React.Component {
  state = {
    displayTableMenu: false,
    menuHeaders: [],
    windowWidth: ''
  }

  _breakpoint = 767;

  _scrollPixels = 200;

  _timeOutDelay = 500;

  componentDidMount = () => {
    this.menuHeadersOnMobile();
    window.addEventListener( 'click', this.toggleTableMenu );
    window.addEventListener( 'keydown', this.handleKbdAccess );
    window.addEventListener( 'resize', this.menuHeadersOnResize );
  }

  componentDidUpdate = ( _, prevState ) => {
    if ( !prevState.displayTableMenu && this.state.displayTableMenu ) {
      const columns = this.getColumns();
      this.handleCheckboxFocus( columns, 0 );
    }
  }

  componentWillUnmount = () => {
    window.removeEventListener( 'click', this.toggleTableMenu );
    window.removeEventListener( 'keydown', this.handleKbdAccess );
    window.removeEventListener( 'resize', this.menuHeadersOnResize );
  }

  setRef = ( node, ref ) => {
    this[ref] = node;
  }

  getColumns = () => (
    this.props.columnMenu.reduce( ( acc, cur ) => (
      [...acc, cur.label]
    ), [] )
  )

  handleCheckboxFocus = ( array, i ) => {
    this[array[i]].inputRef.focus();
  };

  handleCloseMenu = () => this.setState( { displayTableMenu: false } );

  handleKbdAccess = e => {
    if ( !this.state.displayTableMenu ) return;

    if ( ['Home', 'End', 'ArrowDown', 'ArrowUp'].indexOf( e.key ) > -1 ) {
      e.preventDefault();
    }

    const columns = this.getColumns();
    const current = columns.indexOf( e.target.id );
    const first = 0;
    const last = columns.length - 1;
    const next = current + 1;
    const previous = current - 1;
    const isCheckbox = e.target.type === 'checkbox';

    let i;
    switch ( e.key ) {
      case 'Escape':
        this.handleCloseMenu();
        break;

      case 'Tab':
        if ( isCheckbox ) this.handleCloseMenu();
        break;

      case 'ArrowDown':
        i = current === last ? first : next;
        this.handleCheckboxFocus( columns, i );
        break;

      case 'ArrowUp':
        i = current === first ? last : previous;
        this.handleCheckboxFocus( columns, i );
        break;

      case 'Home':
        this.handleCheckboxFocus( columns, first );
        break;

      case 'End':
        this.handleCheckboxFocus( columns, last );
        break;

      default:
        break;
    }
  }

  handleTableScroll = e => {
    const itemsTable = document.querySelector( '.items_table' );
    const tableArrowDirection = e.target.dataset.tablearrow;
    if ( tableArrowDirection === 'right' ) {
      itemsTable.scrollLeft += this._scrollPixels;
    } else {
      itemsTable.scrollLeft -= this._scrollPixels;
    }
  }

  toggleTableMenu = e => {
    const isTableMenu = !!e.target.dataset.tablemenu;
    const isTableMenuItem = ( e.target.parentNode && !!e.target.parentNode.dataset.tablemenuitem ) || !!e.target.dataset.tablemenuitem;
    const isShowMoreColumns = e.target.id === 'show-more-columns';

    this.setState( prevState => {
      if ( isTableMenu ) {
        return { displayTableMenu: !prevState.displayTableMenu };
      }

      if ( isTableMenuItem || isShowMoreColumns ) {
        return { displayTableMenu: true };
      }

      return { displayTableMenu: false };
    } );
  }

  toggleCheckbox = ( e, data ) => {
    const selectedCheckbox = data[`data-proplabel`];
    const { menuHeaders } = this.state;

    this.setState( prevState => {
      if ( menuHeaders.includes( selectedCheckbox ) ) {
        return {
          menuHeaders: prevState.menuHeaders.filter( header => header !== selectedCheckbox )
        };
      }
      return {
        menuHeaders: [...prevState.menuHeaders, selectedCheckbox]
      };
    } );
  }

  menuHeadersOnMobile = () => {
    const allMenuHeaders = this.props.columnMenu.map( menu => menu.label );
    if ( isMobile() ) {
      this.setState( {
        menuHeaders: [...allMenuHeaders]
      } );
    }
  }

  menuHeadersOnResize = () => {
    const windowWidth = window.innerWidth;
    const prevWindowWidth = this.state.windowWidth;

    let resizeMenuHeadersTimer = null;
    if ( resizeMenuHeadersTimer !== null ) clearTimeout( resizeMenuHeadersTimer );
    resizeMenuHeadersTimer = setTimeout( () => {
      if ( prevWindowWidth !== '' && prevWindowWidth <= this._breakpoint && !isWindowWidthLessThanOrEqualTo( this._breakpoint ) ) {
        return this.setState( { menuHeaders: [], windowWidth } );
      }
      if ( isWindowWidthLessThanOrEqualTo( this._breakpoint ) ) {
        const allMenuHeaders = this.props.columnMenu.map( menu => menu.label );
        return this.setState( { menuHeaders: [...allMenuHeaders], windowWidth } );
      }
      return this.setState( { windowWidth } );
    }, this._timeOutDelay );
  }

  render() {
    const { displayTableMenu, menuHeaders } = this.state;
    const { columnMenu, tableMenuOnChange } = this.props;
    const isTableScrollable = menuHeaders.length >= 2;

    return (
      <div className="items_menu_wrapper">
        <div className={ displayTableMenu ? 'items_menu active' : 'items_menu' }>
          <Accordion as={ Menu } vertical>
            <Menu.Item>
              <Accordion.Title
                active={ !displayTableMenu }
                aria-controls="show-more-columns"
                aria-expanded={ displayTableMenu }
                aria-haspopup
                as="button"
                data-tablemenu
                id="show-more-btn"
              >
                Show More <VisuallyHidden el="span">columns</VisuallyHidden>
                <Icon
                  data-tablemenu
                  name={ `angle ${displayTableMenu ? 'up' : 'down'}` }
                />
              </Accordion.Title>
              { displayTableMenu
                && (
                  <Accordion.Content
                    active={ displayTableMenu }
                    aria-hidden={ !displayTableMenu }
                    aria-labelledby="show-more-btn"
                    as="ul"
                    id="show-more-columns"
                    role="menu"
                  >
                    { columnMenu.map( item => (
                      <li key={ item.name }>
                        <Checkbox
                          checked={ menuHeaders.includes( item.label ) }
                          data-proplabel={ item.label }
                          data-propname={ item.name }
                          data-tablemenuitem
                          id={ item.label }
                          label={ titleCase( item.label ) }
                          onChange={ tableMenuOnChange }
                          onClick={ this.toggleCheckbox }
                          ref={ node => this.setRef( node, item.label ) }
                          role="menuitem"
                          tabIndex="-1"
                        />
                      </li>
                    ) ) }
                  </Accordion.Content>
                ) }
            </Menu.Item>
          </Accordion>

          <button
            data-tablearrow="left"
            onClick={ this.handleTableScroll }
            onFocus={ displayTableMenu ? this.toggleTableMenu : null }
            type="button"
            disabled={ !isTableScrollable }
          >
            <VisuallyHidden el="span">scroll table left</VisuallyHidden>
            <Icon name="angle left" data-tablearrow="left" />
          </button>
          <button
            data-tablearrow="right"
            onClick={ this.handleTableScroll }
            onFocus={ displayTableMenu ? this.toggleTableMenu : null }
            type="button"
            disabled={ !isTableScrollable }
          >
            <VisuallyHidden el="span">scroll table right</VisuallyHidden>
            <Icon name="angle right" data-tablearrow="right" />
          </button>
        </div>
      </div>
    );
  }
}

TableMenu.propTypes = {
  columnMenu: PropTypes.array,
  tableMenuOnChange: PropTypes.func
};

TableMenu.defaultProps = {
  columnMenu: []
};

export default TableMenu;
