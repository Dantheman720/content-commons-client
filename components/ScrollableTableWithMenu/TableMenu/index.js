/**
 *
 * TableMenu
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import {
  Accordion, Checkbox, Icon, Menu
} from 'semantic-ui-react';
import VisuallyHidden from 'components/VisuallyHidden/VisuallyHidden';
import { titleCase } from 'lib/utils';
import { isMobile, isWindowWidthLessThanOrEqualTo } from 'lib/browser';
import './TableMenu.scss';

class TableMenu extends React.Component {
  state = {
    displayTableMenu: false,
    menuHeaders: [],
    windowWidth: ''
  }

  breakpoint = 767;

  componentDidMount = () => {
    this.menuHeadersOnMobile();
    window.addEventListener( 'keyup', this.handleKbdAccess );
    window.addEventListener( 'resize', this.menuHeadersOnResize );
  }

  componentWillUnmount = () => {
    window.removeEventListener( 'keyup', this.handleKbdAccess );
    window.removeEventListener( 'resize', this.menuHeadersOnResize );
  }

  menuHeadersOnMobile = () => {
    const allMenuHeaders = this.props.columnMenu.map( menu => menu.label );
    if ( isMobile() ) {
      this.setState( {
        menuHeaders: [...allMenuHeaders]
      } );
    }
  }

  menuHeadersOnResize = () => {
    const windowWidth = window.innerWidth;
    const prevWindowWidth = this.state.windowWidth;

    let resizeMenuHeadersTimer = null;
    if ( resizeMenuHeadersTimer !== null ) clearTimeout( resizeMenuHeadersTimer );
    resizeMenuHeadersTimer = setTimeout( () => {
      if ( prevWindowWidth !== '' && prevWindowWidth <= this.breakpoint && !isWindowWidthLessThanOrEqualTo( this.breakpoint ) ) {
        return this.setState( { menuHeaders: [], windowWidth } );
      }
      if ( isWindowWidthLessThanOrEqualTo( this.breakpoint ) ) {
        const allMenuHeaders = this.props.columnMenu.map( menu => menu.label );
        return this.setState( { menuHeaders: [...allMenuHeaders], windowWidth } );
      }
      return this.setState( { windowWidth } );
    }, 500 );
  }

  handleKbdAccess = e => {
    const isTableMenu = e.target.dataset.tablemenu;
    const isItemsPerPage = e.target.id === 'items-per-page';

    if ( e.key === 'Escape' || ( isTableMenu && e.key === 'Shift' ) || ( isItemsPerPage && e.key === 'Shift' ) ) {
      this.setState( { displayTableMenu: false } );
    }
  }

  toggleTableMenu = () => {
    this.setState( prevState => ( {
      displayTableMenu: !prevState.displayTableMenu
    } ) );
  }

  toggleCheckbox = ( e, data ) => {
    const selectedCheckbox = data[`data-proplabel`];
    const { menuHeaders } = this.state;

    this.setState( prevState => {
      if ( menuHeaders.includes( selectedCheckbox ) ) {
        return {
          menuHeaders: prevState.menuHeaders.filter( header => header !== selectedCheckbox )
        };
      }
      return {
        menuHeaders: [...prevState.menuHeaders, selectedCheckbox]
      };
    } );
  }

  handleTableScroll = e => {
    const itemsTable = document.querySelector( '.items_table' );
    const tableArrowDirection = e.target.dataset.tablearrow;
    if ( tableArrowDirection === 'right' ) {
      itemsTable.scrollLeft += 200;
    } else {
      itemsTable.scrollLeft -= 200;
    }
  }

  render() {
    const { displayTableMenu, menuHeaders } = this.state;
    const { columnMenu, tableMenuOnChange } = this.props;

    return (
      <div className="items_menu_wrapper">
        <div className={ displayTableMenu ? 'items_menu active' : 'items_menu' }>
          <Accordion as={ Menu } vertical>
            <Menu.Item>
              <Accordion.Title
                active={ !displayTableMenu }
                aria-controls="show-more-columns"
                aria-expanded={ displayTableMenu }
                aria-haspopup
                as="button"
                data-tablemenu
                id="show-more-btn"
                onClick={ this.toggleTableMenu }
              >
                Show More <VisuallyHidden el="span">columns</VisuallyHidden>
                <Icon name={ `angle ${displayTableMenu ? 'up' : 'down'}` } />
              </Accordion.Title>
              { displayTableMenu
                && (
                  <Accordion.Content
                    active={ displayTableMenu }
                    aria-hidden={ !displayTableMenu }
                    aria-labelledby="show-more-btn"
                    as="ul"
                    id="show-more-columns"
                    role="menu"
                  >
                    { columnMenu.map( item => (
                      <li key={ item.name } role="menuitem">
                        <Checkbox
                          checked={ menuHeaders.includes( item.label ) }
                          data-proplabel={ item.label }
                          data-propname={ item.name }
                          data-tablemenuitem
                          id={ item.label }
                          label={ titleCase( item.label ) }
                          onChange={ tableMenuOnChange }
                          onClick={ this.toggleCheckbox }
                        />
                      </li>
                    ) ) }
                  </Accordion.Content>
                ) }
            </Menu.Item>
          </Accordion>

          <button
            data-tablearrow="left"
            onClick={ this.handleTableScroll }
            onFocus={ displayTableMenu ? this.toggleTableMenu : null }
            type="button"
          >
            <VisuallyHidden el="span">scroll table left</VisuallyHidden>
            <Icon name="angle left" data-tablearrow="left" />
          </button>
          <button
            data-tablearrow="right"
            onClick={ this.handleTableScroll }
            onFocus={ displayTableMenu ? this.toggleTableMenu : null }
            type="button"
          >
            <VisuallyHidden el="span">scroll table right</VisuallyHidden>
            <Icon name="angle right" data-tablearrow="right" />
          </button>
        </div>
      </div>
    );
  }
}

TableMenu.propTypes = {
  columnMenu: PropTypes.array,
  tableMenuOnChange: PropTypes.func
};

export default TableMenu;
