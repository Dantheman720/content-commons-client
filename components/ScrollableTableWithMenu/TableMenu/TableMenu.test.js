import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import TableMenu from './TableMenu';

const props = {
  columnMenu: [
    { name: 'author', label: 'AUTHOR' },
    { name: 'categories', label: 'CATEGORIES' },
    { name: 'createdAt', label: 'DATE' }
  ],
  tableMenuOnChange: jest.fn()
};

const Component = <TableMenu { ...props } />;

describe( '<TableMenu />', () => {
  it( 'renders without crashing', () => {
    const wrapper = shallow( Component );

    expect( wrapper.exists() ).toEqual( true );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  it( 'componentDidMount calls menuHeadersOnMobile and ', () => {
    const wrapper = shallow( Component );
    const inst = wrapper.instance();
    const spy = jest.spyOn( inst, 'menuHeadersOnMobile' );

    inst.componentDidMount();
    expect( spy ).toHaveBeenCalled();
  } );

  it( 'componentDidMount attaches event listeners', () => {
    const wrapper = shallow( Component );
    const inst = wrapper.instance();
    const clickSpy = jest.spyOn( inst, 'toggleTableMenu' )
      .mockImplementation( () => {} );
    const keydownSpy = jest.spyOn( inst, 'handleKbdAccess' )
      .mockImplementation( () => {} );
    const resizeSpy = jest.spyOn( inst, 'menuHeadersOnResize' )
      .mockImplementation( () => {} );

    const map = {};
    window.addEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    inst.componentDidMount();

    map.click();
    expect( clickSpy ).toHaveBeenCalled();

    map.keydown();
    expect( keydownSpy ).toHaveBeenCalled();

    map.resize();
    expect( resizeSpy ).toHaveBeenCalled();
  } );

  it( 'componentWillUnmount removes event listeners', () => {
    const wrapper = shallow( Component );
    const inst = wrapper.instance();
    const clickSpy = jest.spyOn( inst, 'toggleTableMenu' )
      .mockImplementation( () => {} );
    const keydownSpy = jest.spyOn( inst, 'handleKbdAccess' )
      .mockImplementation( () => {} );
    const resizeSpy = jest.spyOn( inst, 'menuHeadersOnResize' )
      .mockImplementation( () => {} );

    const map = {};
    window.removeEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    Promise.resolve()
      .then( () => {
        inst.componentWillUnmount();
        expect( clickSpy ).toHaveBeenCalled();
        expect( keydownSpy ).toHaveBeenCalled();
        expect( resizeSpy ).toHaveBeenCalled();
      } )
      .catch( () => {} );
  } );

  it( 'getColumns returns an array of columnMenu labels', () => {
    const wrapper = shallow( Component );
    const inst = wrapper.instance();
    const columns = inst.getColumns();

    columns.forEach( ( column, i ) => {
      expect( column ).toEqual( props.columnMenu[i].label );
    } );
  } );

  it( 'componentDidUpdate calls getColumns and handleCheckboxFocus', () => {
    const wrapper = shallow( Component );
    const inst = wrapper.instance();
    const columnsSpy = jest.spyOn( inst, 'getColumns' )
      .mockImplementation( () => {} );
    const checkboxSpy = jest.spyOn( inst, 'handleCheckboxFocus' )
      .mockImplementation( () => {} );

    // check initial state
    expect( inst.state.displayTableMenu ).toEqual( false );

    wrapper.setState( { displayTableMenu: true } );
    expect( inst.state.displayTableMenu ).toEqual( true );
    expect( columnsSpy ).toHaveBeenCalled();
    expect( checkboxSpy ).toHaveBeenCalled();
  } );

  it( 'handleCloseMenu sets displayTableMenu state to false', () => {
    const wrapper = shallow( Component );
    const inst = wrapper.instance();
    jest.spyOn( inst, 'componentDidUpdate' )
      .mockImplementation( () => {} );

    // set to true first
    wrapper.setState( { displayTableMenu: true } );

    inst.handleCloseMenu();
    expect( inst.state.displayTableMenu ).toEqual( false );
  } );

  it( 'setRef assigns a ref with a name', () => {
    const wrapper = shallow( Component );
    const inst = wrapper.instance();
    jest.spyOn( inst, 'handleCheckboxFocus' )
      .mockImplementation( () => {} );
    const node = <div>test node</div>;
    const refName = 'AUTHOR';

    wrapper.setState( { displayTableMenu: true } );
    inst.setRef( node, refName );

    expect( inst[refName].type ).toEqual( 'div' );
    expect( inst[refName].props.children ).toEqual( 'test node' );
  } );
} );
