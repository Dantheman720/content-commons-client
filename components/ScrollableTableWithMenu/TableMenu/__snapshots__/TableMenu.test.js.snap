// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableMenu /> renders without crashing 1`] = `
<div
  className="items_menu_wrapper"
>
  <div
    className="items_menu"
  >
    <Accordion
      as={[Function]}
      vertical={true}
    >
      <MenuItem>
        <AccordionTitle
          active={true}
          aria-controls="show-more-columns"
          aria-expanded={false}
          aria-haspopup={true}
          as="button"
          data-tablemenu={true}
          id="show-more-btn"
        >
          Show More 
          <VisuallyHidden
            el="span"
          >
            columns
          </VisuallyHidden>
          <Icon
            as="i"
            data-tablemenu={true}
            name="angle down"
          />
        </AccordionTitle>
      </MenuItem>
    </Accordion>
    <button
      data-tablearrow="left"
      disabled={true}
      onClick={[Function]}
      onFocus={null}
      type="button"
    >
      <VisuallyHidden
        el="span"
      >
        scroll table left
      </VisuallyHidden>
      <Icon
        as="i"
        data-tablearrow="left"
        name="angle left"
      />
    </button>
    <button
      data-tablearrow="right"
      disabled={true}
      onClick={[Function]}
      onFocus={null}
      type="button"
    >
      <VisuallyHidden
        el="span"
      >
        scroll table right
      </VisuallyHidden>
      <Icon
        as="i"
        data-tablearrow="right"
        name="angle right"
      />
    </button>
  </div>
</div>
`;
